core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyBuilder.java

```diff
@@ -0,0 +1,68 @@
+package org.springframework.security.access.hierarchicalroles;
+
+import org.springframework.util.Assert;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.springframework.security.access.hierarchicalroles.RoleHierarchyUtils.roleHierarchyFromMap;
+
+/**
+ * Class to provide a fluent API to build a {@link String} representing a {@link RoleHierarchy}.
+ * Relies on  {@link RoleHierarchyUtils#roleHierarchyFromMap(Map)} to build the string.
+ */
+public class RoleHierarchyBuilder {
+	private final Map<String, List<String>> roleBranches;
+
+	private RoleHierarchyBuilder() {
+		this.roleBranches = new LinkedHashMap<>();
+	}
+
+	public static RoleHierarchyBuilder newBuilder() {
```

Comment:
Since any role hierarchy will require at least one role, I think it would be nice to call this `role`, removing one call.

Reply:
Note that I believe it is moved into `RoleHierarchyImpl`, then the two `role` methods won't conflict since this one will live in `RoleHierarchyImpl` itself.

Reply:
Hello @jzheaux 

While implementing this suggestion I think placing the call to the builder in the implementation is kind of mixing the roles of the entities.

I suggest, instead, leave the building of the `RoleHierarchyImpl` to live only in the builder, and refactor the final `build()` method to create a fully featured `RoleHierarchyImpl` (using the utility method `RoleHierarchyUtills#roleHierarchyFromMap` and `RoleHierarchyImpl#setHierarchy`, this way, it will have a complete and clearer scope and will reduce the impact on related classes, it will also allow doing something like the follow:

```
@Bean
public RoleHierarchy roleHierarchy() {
    return RoleHierarchyBuilder.newBuilder()
        .role("ROLE_A").impliesAllOf("ROLE_B")
        .role("ROLE_B").impliesAllOf("ROLE_C", "ROLE_D")
        .build();
}
```

Elegant and concise, really straightforward. 
Wdyt?

Reply:
Thank you for your feedback and help to further understand your idea, @jzheaux !

Makes total sense so I'm planning to move the builder to the class `RoleHierarchyImpl` to follow the convention of the classes you mention.

Will try to implement  a `role(String)` method in that class to better align with the last comment in this thread.

Thank you for sharing your thoughts!